### DISCUSSION TITLE
Missing type of services in ApiExtensionContext

### DISCUSSION THREAD

**Discussion by Zehir** [directus/directus#24870] [Feature Requests]
### Describe the Improvement

Hello,

Long story short, it's started from `defineHook`, in the provided context the property `services` from which one we get the complete app is missing his type.

The `any` is located here :
https://github.com/directus/directus/blob/d50e52f9e9990762ac1c3a19a3ba6c66dbaf7f70/packages/types/src/extensions.ts#L87-L93

In my extension I was able to load the type definition with :
```typescript
import type * as Services from '@directus/api/dist/services/index'

export default defineHook(async ({ action }, ctx) => {
  const services = ctx.services as typeof Services
})
```
Edit : Work in editor but can't build it

Even if it's working it's confusing me because of the package.json definition;

https://github.com/directus/directus/blob/d50e52f9e9990762ac1c3a19a3ba6c66dbaf7f70/api/package.json#L50-L55

I don't understand why the  `/dist` is required.
I tried to do a PR to fix this but I don't know enought the codebase to do it. I tried to import it in the `@directus/types` packages but got no success with it.

Thanks

**Comment by Zehir** [comment]
From https://github.com/directus/directus/pull/19810#issuecomment-1734585917
> @Zehir The problem with the proposed solution is that it creates a circular dependency between API and `@directus/types` 😬 My preferred solution would be to split off the services into a `@directus/services` and then have that imported in both `/types` and `/api`

The problem is that there is actives PR so a PR to create the `@directus/services` packages would broke some PR.
Maybe wait after the market places PRs ?

If it's only moving the file and update the import It's could take time but not that hard. I can take an existing package as example like `@directus/stores` or it's more complex than that ? Maybe there is cross import that screw this up. Correct me if I am wrong but I think the export of the `@directus/services` packages will only contain definition for every services.

Side question, I was wondering why there is package at the root level and some in the `packages` directory, Is that because of the transition to splited packages ?

**Comment by rijkvanzanten** [comment]
> Correct me if I am wrong but I think the export of the @directus/services packages will only contain definition for every services.

My longer term goal is to have the actual services themselves explicitly outside of `/api` so we can also split off the usage for `/cli` out of the api codebase to keep the separation of concerns intact. The `/api` and `/app` packages respectively got way too big for my taste 😬 

> Side question, I was wondering why there is package at the root level and some in the packages directory, Is that because of the transition to splited packages ?

Great question! It's a bit of that, and a little bit of that being the packages that most people have the access the most often

**Comment by Xyz3R** [comment]
Is this being worked on? Missing types for services in extensions is kinda annoying

**Comment by lu40** [comment]
Workaround for this issue:

1. Add `@directus/api`:
```bash
npm i -D @directus/api
```

2. Import this and create a typed version of the services object:
```typescript
import { defineOperationApi } from '@directus/extensions-sdk';
import * as Services from '@directus/api/dist/services';

type Post = {
  title: string;
};

export default defineOperationApi({
  id: 'demo',
  handler: async (_options, { services, getSchema }) => {
    const schema = await getSchema();

    const typedServices = services as typeof Services;

    const { ItemsService } = typedServices;

    const postsService = new ItemsService<Post, 'posts'>('posts', {
      schema,
    });

    const postId = await postsService.createOne({
      title: 'Hello World!',
    });

    // postId: PrimaryKey
  },
});

```

**Comment by TBG-FR** [comment]
We really need an improvement like this, to be able to import every directus built-in types when developing an extension without going around the main goal of Typescript !

EDIT : Here's what I ended up doing to be able to use types everywhere
```ts
/**
 * Custom properties on the req object in express
 */
import { Query } from '@directus/shared/types';
import type { SchemaOverview, Accountability } from '@directus/types';
// import type { SchemaOverview, Accountability } from '@directus/shared/types'; => errors

declare global {
	namespace Express {
		export interface Request {
			token: string | null;
			collection: string;
			sanitizedQuery: Query;
			schema: SchemaOverview;

			accountability?: Accountability;
			singleton?: boolean;
		}
	}
}
```
```ts
export default
{
	id: 'fourmis',
	handler: (router:Router, context:ApiExtensionContext) =>
	{
		const { getSchema, env, logger } = context;
		const typedServices = context.services as typeof Services;
		// FieldsService: Provides access to perform operations on fields used in collections and items.
		const { 
			ItemsService, CollectionsService, FieldsService, FilesService,
			UsersService, RolesService, PermissionsService
		} = typedServices;

		router.get('/', (req:Request, res:Response) => 
		{
		  /* code here */
		}
	},
};
```

**Comment by pinkdose** [comment]
@directus/api/dist/services types should really be integrated with the main built-in types that comes together with the extension library

**Comment by that1matt** [comment]
This has also caused me some trouble, so I created a PR to hopefully provide better TS support, and provide intellisense for the Services. This doesn't cover all services, but it should help. Any feedback is always appreciated.


See https://github.com/directus/directus/pull/25368

**Reply by that1matt** [reply]
The changes in the PR have been merged in, and will be available in the next release. The types should include all services available, and make for a better DX within the APIExtensionContext.

**Reply by rijkvanzanten** [reply]
🙌🙌🙌🙌🙌
